File: internal/radio/stations.go
package radio

import (
	_ "embed"
	"encoding/csv"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/agejevasv/goradion/internal/logging"
)

//go:embed stations.csv
var defaultStationsCSV string

type Station struct {
	title string
	url   string
	tags  []string
}

// LoadStations reads the stations CSV from a file or an http(s) URL; an empty
// source means the built-in list, which also stands in for a URL that can't
// be fetched.
func LoadStations(source string) ([]Station, error) {
	switch {
	case source == "":
		return parseStations(strings.NewReader(defaultStationsCSV))
	case strings.HasPrefix(source, "http://"), strings.HasPrefix(source, "https://"):
		body, err := fetchStations(source)
		if err != nil {
			logging.Printf("stations: %v, using the built-in list", err)
			return parseStations(strings.NewReader(defaultStationsCSV))
		}
		defer body.Close()
		return parseStations(body)
	default:
		f, err := os.Open(source)
		if err != nil {
			return nil, err
		}
		defer f.Close()
		return parseStations(f)
	}
}

// parseStations reads "title,url[,tag;tag...]" rows and skips rows without a
// URL.
func parseStations(r io.Reader) ([]Station, error) {
	reader := csv.NewReader(r)
	reader.FieldsPerRecord = -1
	records, err := reader.ReadAll()
	if err != nil {
		return nil, fmt.Errorf("can't parse stations CSV: %w", err)
	}

	stations := make([]Station, 0, len(records))
	for _, rec := range records {
		if len(rec) < 2 || strings.TrimSpace(rec[1]) == "" {
			continue
		}
		s := Station{title: strings.TrimSpace(rec[0]), url: strings.TrimSpace(rec[1])}
		if len(rec) > 2 {
			for _, tag := range strings.Split(rec[2], ";") {
				if tag = strings.TrimSpace(tag); tag != "" {
					s.tags = append(s.tags, tag)
				}
			}
		}
		stations = append(stations, s)
	}
	return stations, nil
}

func fetchStations(url string) (io.ReadCloser, error) {
	client := &http.Client{Timeout: 15 * time.Second}
	resp, err := client.Get(url)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode != http.StatusOK {
		resp.Body.Close()
		return nil, fmt.Errorf("GET %s: status %d", url, resp.StatusCode)
	}
	return resp.Body, nil
}

File: rust/src/stations.rs
use std::io::Read;

use crate::audio::http;
use crate::log;

const BUILT_IN: &str = include_str!("../stations.csv");

#[derive(Clone, Debug, PartialEq)]
pub struct Station {
    pub title: String,
    pub url: String,
    pub tags: Vec<String>,
}

/// Reads the stations CSV from a file or an http(s) URL; an empty source means
/// the built-in list, which also stands in for a URL that can't be fetched.
pub fn load(source: &str) -> Result<Vec<Station>, String> {
    if source.is_empty() {
        return parse(BUILT_IN.as_bytes());
    }
    if source.starts_with("http://") || source.starts_with("https://") {
        let mut body = Vec::new();
        match http::get(source, &[]).and_then(|mut r| r.body.read_to_end(&mut body)) {
            Ok(_) => return parse(&body[..]),
            Err(e) => {
                log!("stations: {e}, using the built-in list");
                return parse(BUILT_IN.as_bytes());
            }
        }
    }
    let file = std::fs::File::open(source).map_err(|e| format!("{source}: {e}"))?;
    parse(file)
}

/// Reads "title,url[,tag;tag...]" rows and skips rows without a URL.
fn parse(r: impl Read) -> Result<Vec<Station>, String> {
    let mut reader = csv::ReaderBuilder::new().has_headers(false).flexible(true).from_reader(r);
    let mut stations = Vec::new();
    for rec in reader.records() {
        let rec = rec.map_err(|e| format!("can't parse stations CSV: {e}"))?;
        let url = rec.get(1).unwrap_or("").trim();
        if url.is_empty() {
            continue;
        }
        let tags =
            rec.get(2).unwrap_or("").split(';').map(str::trim).filter(|t| !t.is_empty()).map(String::from).collect();
        stations.push(Station { title: rec[0].trim().to_string(), url: url.to_string(), tags });
    }
    Ok(stations)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_rows() {
        let csv = "A,http://a,Jazz; Lounge\n\"B, the station\",http://b\nno url\nC,,Rock\n";
        let got = parse(csv.as_bytes()).unwrap();
        assert_eq!(got.len(), 2);
        assert_eq!(got[0].tags, ["Jazz", "Lounge"]);
        assert_eq!(got[1].title, "B, the station");
        assert!(got[1].tags.is_empty());
    }

    #[test]
    fn built_in_list() {
        assert!(load("").unwrap().len() > 100);
    }
}
